UI: Add ctrl + scroll to zoom scene items

exeldro opened the pull request:
### Description
Add ctrl + scroll to zoom the selected scene items to the position of the mouse pointer

### Motivation and Context
This makes zooming in to an interesting part much easier. Before you had to resize and move the scene items.

### How Has This Been Tested?
On windows 64 bit with different bounding box types and different positional alignments.

### Types of changes
- New feature (non-breaking change which adds functionality) 

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

tt2468 commented:
Oh wow I love this idea. Does it control whatever source the cursor is hovering over or does it control whatever source is selected?

exeldro commented:
It controls the selected sources. So it can zoom multiple sources at once.

unknowndomain commented:
I've built this on macOS on a MacBook Pro with a Touch Pad, it works but the inertial scrolling isn't the best experience. Obviously using the Command key instead.

I noted that you are using "obsolete" for new code features like `QWheelEvent::delta()` and as a result had to write code to deal with HDPI devices, my understanding is that `pixelDelta()` provides this for you but isn't recommended for X11 devices.

It might be useful to consider adding a setting for controlling the scaling speed, or detecting whether the user has a trackpad with innertial scrolling rather than a mouse wheel.

Alternatively a pinch gesture event could work but without the modifier key?

exeldro commented:
Changed it to use pixelDelta and angleDelta, but I have no way of testing the pixelDelta version.
@unknowndomain can you test this for me?

I think handling the pinch gesture event should be a separate pull request, which I can not make because I don't have a touch pad.

unknowndomain commented:
Weirdly it doesn't work when compiled from the CLI on macOS but it does work from Qt Creator which is building in the same place, no idea what/why but it seems fine.

The scrolling seems to work well, it's slow but that allows more precision than before where a small scroll would be a massive change.

cg2121 commented:
I think there should be something in the transform dialog to control the zoom as well.

PatTheMav commented:
Just rebased this on master and checked it out on macOS - looks fine to me, scrolling felt "right" with a touchpad.

> Alternatively a pinch gesture event could work but without the modifier key?

Pinch gestures would feel "right" on macOS, but I haven't fully wrapped my head around what they should do (i.e. zoom the source vs. zoom the preview), so I'm fine with merging this feature for all platforms as-is for now.

JonGarciaP commented:
where can I have instructions on how to install this? 
I probably don't belong here but I would still like to know :/

JonGarciaP commented:
> where can I have instructions on how to install this?
> I probably don't belong here but I would still like to know :/

I have not found any instructions on how to get this working :(

cg2121 reviewed (changes requested):
Just tested it out and it works great. I would suggest though to have a maximum zoom / zoom out, as the zooming now can go on forever.

Warchamp7 commented:
Agree with all of Matt's feedback here

ryantheleach commented:
I believe that the following issues are asking for very similar related functionality, in that the zoom can be performed whilst allowing the bounding boxes to remain in their existing positions.

It'd be good to be able to kill 2 birds with 1 stone, and set up controls precedent for also allowing panning the same way.

e.g. reverse the controls so that scrolling on a source zooms it, and control makes it respect it's existing bounds while zooming.
Then panning is just clicking and dragging, and panning while respecting bounds is ctrl and dragging.



https://ideas.obsproject.com/posts/1420/shift-alt-crop-towards-center-and-shift-drag-to-pan
https://ideas.obsproject.com/posts/747/easily-pan-a-cropped-source
https://ideas.obsproject.com/posts/1181/add-option-to-crop-to-bounding-box-in-source-transform
https://ideas.obsproject.com/posts/550/virtual-ptz-camera

The last PTZ point, is essentially just animating crops and positioning, which if this became easier to edit in the canvas, would just be able to be handled by move transition plugin or move values for the time being, or using move filters in a nicer GUI.

Morgyn commented:
I hope this gets added soon!

znbaboy commented:
still waiting for this... <3

exeldro commented:
Rebased and added Size All Cursor
![2025-03-1708-56-54-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/863fe03f-feeb-4511-af57-336440427185)

PatTheMav commented:
@Warchamp7 Please give this a review so we can possibly schedule a possible release for it if we still think it's a valuable change.
